File: Source/Zybach.API/Services/GeoOptixService.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Threading.Tasks;
using GeoJSON.Net.Feature;
using GeoJSON.Net.Geometry;
using Microsoft.Extensions.Logging;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using Zybach.Models.DataTransferObjects;
using Zybach.Models.GeoOptix;
using Zybach.Models.GeoOptix.SampleMethod;
using JsonSerializer = Newtonsoft.Json.JsonSerializer;

namespace Zybach.API.Services
{
    public class GeoOptixService
    {
        private const string GeoOptixSitesProjectOverviewWebUri = "project-overview-web/water-data-program/sites";
        private const string GeoOptixSitesProjectsUri = "projects/water-data-program/sites";

        private readonly HttpClient _httpClient;
        private readonly ILogger<GeoOptixService> _logger;

        public GeoOptixService(HttpClient httpClient, ILogger<GeoOptixService> logger)
        {
            _httpClient = httpClient;
            _logger = logger;

        }

        private async Task<TV> GetJsonFromCatalogImpl<TV>(string uri)
        {
            using var httpResponse = await _httpClient.GetAsync(uri, HttpCompletionOption.ResponseHeadersRead);
            httpResponse.EnsureSuccessStatusCode(); // throws if not 200-299
            using var streamReader = new StreamReader(httpResponse.Content.ReadAsStreamAsync().Result);
            using var jsonTextReader = new JsonTextReader(streamReader);
            return new JsonSerializer().Deserialize<TV>(jsonTextReader);
        }

        public async Task<List<WellSummaryDto>> GetWellSummaries()
        {
            var geoOptixSites = await GetGeoOptixSites();
            return geoOptixSites.AsParallel().Select(x => new WellSummaryDto(x)).ToList();
        }

        public async Task<List<WellSummaryDto>> GetWellSummariesCreatedAsOfYear(int year)
        {
            var geoOptixSites = await GetGeoOptixSites();
            return geoOptixSites.Where(x => x.CreateDate.Year <= year).AsParallel().Select(x => new WellSummaryDto(x)).ToList();
        }

        public async Task<AbbreviatedWellDataResponse> GetWellAsAbbreviatedWellDataResponse(string wellRegistrationID)
        {
            var geoOptixSite = await GetGeoOptixSite(wellRegistrationID);
            var geoOptixStations = await GetGeoOptixStationsForSite(wellRegistrationID);
            return new AbbreviatedWellDataResponse(geoOptixSite, geoOptixStations);
        }

        public async Task<WellSummaryDto> GetWellSummary(string wellRegistrationID)
        {
            var geoOptixSite = await GetGeoOptixSite(wellRegistrationID);
            return new WellSummaryDto(geoOptixSite);
        }

        private async Task<List<Site>> GetGeoOptixSites()
        {
            var geoOptixSites = await GetJsonFromCatalogImpl<List<Site>>(GeoOptixSitesProjectOverviewWebUri);
            return geoOptixSites;
        }

        public async Task<Site> GetGeoOptixSite(string siteID)
        {
            var geoOptixSite = await GetJsonFromCatalogImpl<Site>($"{GeoOptixSitesProjectOverviewWebUri}/{siteID}");
            return geoOptixSite;
        }

        public async Task<List<SensorSummaryDto>> GetSensorSummaries()
        {
            var geoOptixStations = await GetGeoOptixStations();
            return geoOptixStations.Select(x => new SensorSummaryDto(x)).ToList();
        }

        public async Task<List<SensorSummaryDto>> GetSensorSummariesCreatedAsOfYear(int year)
        {
            var geoOptixStations = await GetGeoOptixStations();
            return geoOptixStations.Where(x => x.CreateDate.Year <= year).AsParallel().Select(x => new SensorSummaryDto(x)).ToList();
        }

        public async Task<List<SensorSummaryDto>> GetSensorSummariesForWell(string wellRegistrationID)
        {
            var geoOptixStations = await GetGeoOptixStationsForSite(wellRegistrationID);
            return geoOptixStations.AsParallel().Select(x => new SensorSummaryDto(x)).ToList();
        }

        private async Task<List<Station>> GetGeoOptixStations()
        {
            const string uri = "project-overview-web/water-data-program/stations";
            var geoOptixStations = await GetJsonFromCatalogImpl<List<Station>>(uri);
            return geoOptixStations;
        }

        public async Task<List<Station>> GetGeoOptixStationsForSite(string siteID)
        {
            var geoOptixStations = await GetJsonFromCatalogImpl<List<Station>>($"{GeoOptixSitesProjectOverviewWebUri}/{siteID}/stations");
            return geoOptixStations;
        }

        public async Task<List<Sample>> GetGeoOptixSamplesForSite(string siteID)
        {
            var getGeoOptixSamplesForSite = await GetJsonFromCatalogImpl<List<Sample>>($"{GeoOptixSitesProjectsUri}/{siteID}/samples");
            return getGeoOptixSamplesForSite;
        }

        public async Task<List<SampleMethodDto>> GetGeoOptixSampleMethodsForSite(string siteID, string methodID)
        {
            var geoOptixStations = await GetJsonFromCatalogImpl<List<SampleMethodDto>>($"{GeoOptixSitesProjectsUri}/{siteID}/samples/{methodID}/methods");
            return geoOptixStations;
        }

        public async Task<List<InstallationRecordDto>> GetInstallationRecords(string wellRegistrationID)
        {
            var installationRecordDtos = new List<InstallationRecordDto>();
            var geoOptixSamples = await GetGeoOptixSamplesForSite(wellRegistrationID);
            foreach (var geoOptixSample in geoOptixSamples)
            {
                var installationRecordDto = new InstallationRecordDto();
                var installationCanonicalName = geoOptixSample.CanonicalName;
                var geoOptixSampleMethodsForSite = await GetGeoOptixSampleMethodsForSite(wellRegistrationID, installationCanonicalName);
                if (!geoOptixSampleMethodsForSite.Any())
                {
                    break;
                }

                var geoOptixMethod = geoOptixSampleMethodsForSite.First();

                if (geoOptixMethod.MethodSchemaCanonicalName != "well-install")
                {
                    break;
                }

                var installationRecord = GetRecordInstance(geoOptixMethod.MethodInstance.RecordSets.First());
                var sensorRecord = GetRecordInstance(installationRecord, "sensor");
                var photoRecords = GetRecordInstance(sensorRecord, "photos");
                var sensorType = GetRecordSetValueAsString(sensorRecord, "sensor-type");
                var continuitySensorModelRawValue = installationRecord.Fields.SingleOrDefault(x => x.CanonicalName == "sensor-model-continuity");
                var continuitySensorModel = continuitySensorModelRawValue != null && continuitySensorModelRawValue.RawValue is JArray ? continuitySensorModelRawValue.RawValue.ToObject<List<string>>().First() : GetRecordSetValueAsString(sensorRecord, "sensor-model-flow");
                var flowSensorModel = GetRecordSetValueAsString(sensorRecord, "sensor-model-flow");
                var pressureSensorModel = GetRecordSetValueAsString(sensorRecord, "sensor-model-pressure");

                installationRecordDto.InstallationCanonicalName = installationCanonicalName;
                installationRecordDto.Status = geoOptixMethod.Status.Name;
                installationRecordDto.Date = GetRecordSetValueAsDateTime(installationRecord, "install-date");
                var gpsLocation = GetRecordSetValueAsPoint(installationRecord, "gps-location");
                installationRecordDto.Longitude = gpsLocation?.Coordinates.Longitude;
                installationRecordDto.Latitude = gpsLocation?.Coordinates.Latitude;
                installationRecordDto.FlowMeterSerialNumber = GetRecordSetValueAsString(sensorRecord, "flow-serial-number");
                installationRecordDto.SensorSerialNumber = GetRecordSetValueAsString(sensorRecord, "sensor-serial-number");
                var installerAffiliationRawValue = GetRecordSetValueAsStringList(installationRecord, "installer-affiliation");
                installationRecordDto.Affiliation = installerAffiliationRawValue.Any() ? installerAffiliationRawValue.First().ToUpper() : null;
                installationRecordDto.Initials = GetRecordSetValueAsString(installationRecord, "installer-initials");
                installationRecordDto.SensorType = sensorType;
                installationRecordDto.ContinuitySensorModel = continuitySensorModel;
                installationRecordDto.FlowSensorModel = flowSensorModel;
                installationRecordDto.PressureSensorModel = pressureSensorModel;
                installationRecordDto.WellDepth = GetRecordSetValueAsDouble(sensorRecord, "well-depth");
                installationRecordDto.InstallDepth = GetRecordSetValueAsDouble(sensorRecord, "install-depth");
                installationRecordDto.CableLength = GetRecordSetValueAsDouble(sensorRecord, "cable-length");
                installationRecordDto.WaterLevel = GetRecordSetValueAsDouble(sensorRecord, "well-depth");
                installationRecordDto.Photos = photoRecords.Fields.Where(x => x.CanonicalName == "photo").Select(x => x.RawValue.ToObject<string>()).ToList();
                installationRecordDtos.Add(installationRecordDto);
            }
            return installationRecordDtos;
        }

        public async Task<Stream> GetPhoto(string wellRegistrationID, string installationCanonicalName,
            string photoCanonicalName)
        {
            return await _httpClient.GetStreamAsync(
                $"{GeoOptixSitesProjectsUri}/{wellRegistrationID}/samples/{installationCanonicalName}/folders/.methods/files/{photoCanonicalName}/view");
        }

        private static Point GetRecordSetValueAsPoint(RecordInstance recordInstance, string canonicalName)
        {
            var gpsLocationRawValue = recordInstance.Fields.Single(x => x.CanonicalName == canonicalName).RawValue;
            var gpsLocationAsFeature = gpsLocationRawValue.ToObject<Feature>();
            return gpsLocationAsFeature != null ? (Point) gpsLocationAsFeature.Geometry : null;
        }

        private static RecordInstance GetRecordInstance(RecordInstance recordInstance, string canonicalName)
        {
            var recordSet = GetRecordInstanceFromFieldsByCanonicalName(recordInstance, canonicalName);
            return GetRecordInstance(recordSet);
        }

        private static RecordInstance GetRecordInstance(RecordSet recordSet)
        {
            return recordSet.Records.First().RecordInstance;
        }

        private static RecordSet GetRecordInstanceFromFieldsByCanonicalName(RecordInstance recordInstance, string canonicalName)
        {
            return recordInstance.Fields.Single(x => x.CanonicalName == canonicalName).RawValue.ToObject<RecordSet>();
        }

        private static string GetRecordSetValueAsString(RecordInstance recordInstance, string canonicalName)
        {
            var recordSetValue = recordInstance.Fields.Single(x => x.CanonicalName == canonicalName).RawValue;
            return recordSetValue is JObject 
                ? recordSetValue.ToObject<RecordSet>().Records[0].RecordInstance.Fields[0].RawValue.ToObject<string>() 
                : recordSetValue is JArray? recordSetValue.ToObject<List<string>>()[0] : recordSetValue.ToObject<string>();
        }

        private static List<string> GetRecordSetValueAsStringList(RecordInstance recordInstance, string canonicalName)
        {
            var recordSetValue = recordInstance.Fields.SingleOrDefault(x => x.CanonicalName == canonicalName)?.RawValue;
            return recordSetValue != null ? recordSetValue.ToObject<List<string>>() : new List<string>();
        }

        private static DateTime? GetRecordSetValueAsDateTime(RecordInstance recordInstance, string canonicalName)
        {
            var recordSetValue = recordInstance.Fields.Single(x => x.CanonicalName == canonicalName).RawValue;
            return recordSetValue.ToObject<DateTime?>();
        }

        private static double? GetRecordSetValueAsDouble(RecordInstance recordInstance, string canonicalName)
        {
            var recordSetValue = recordInstance.Fields.Single(x => x.CanonicalName == canonicalName).RawValue;
            return recordSetValue.ToObject<double?>();
        }

        public async Task<Dictionary<string, WellWithSensorSummaryDto>> GetWellsWithSensors()
        {
            var sensors = await GetSensorSummaries();
            var wells = await GetWellSummaries();

            // add a sensor array to the wells
            var wellsWithSensors = wells.Select(x => new WellWithSensorSummaryDto
            {
                WellRegistrationID = x.WellRegistrationID,
                Description = x.Description,
                Location = x.Location,
                Sensors = sensors.Where(y => y.WellRegistrationID == x.WellRegistrationID).ToList(),
                InGeoOptix = true
            }).ToList();

            // create a Map from the array of wells
            return wellsWithSensors.ToDictionary(x => x.WellRegistrationID);
        }
    }
}

File: Source/Zybach.API/src/services/geooptix-service.ts
import axios from "axios";
import { InternalServerError } from "../errors/internal-server-error";
import secrets from "../secrets";
import { provideSingleton } from "../util/provide-singleton";
import { SensorSummaryDto, SensorTypeMap, WellSummaryDto, WellWithSensorSummaryDto } from "../dtos/well-summary-dto";
import { WellController } from "../controllers/wells.controller";
import { GeoOptixObjectTypeToZybachObjectType, SearchSummaryDto } from "../dtos/search-summary-dto";

const FileReader = require('filereader');

@provideSingleton(GeoOptixService)
export class GeoOptixService {
    baseUrl: string;
    searchUrl: string;
    headers: { "x-geooptix-token": string };
    constructor(){
        this.baseUrl = secrets.GEOOPTIX_HOSTNAME;
        this.searchUrl = secrets.GEOOPTIX_SEARCH_HOSTNAME;
        this.headers = {
            "x-geooptix-token": secrets.GEOOPTIX_API_KEY
        }
    }

    public async getWellSummaries(): Promise<WellSummaryDto[]> {
        try {
            // todo: this getting stuff from GeoOptix needs to live in GeoOptixService class
            const geoOptixRequest = await axios.get(
                `${this.baseUrl}/project-overview-web/water-data-program/sites`,
                {
                    headers: this.headers
                }
            );

            return geoOptixRequest.data
                .map((x: { CanonicalName: any; Description: any; Location: any; }) =>
                    ({ wellRegistrationID: x.CanonicalName, description: x.Description, location: x.Location }));
        }
        catch (err) {
            console.error(err);
            throw new InternalServerError(err.message);
        }
    }

    public async getWellSummariesCreatedAsOfYear(year: number): Promise<WellSummaryDto[]> {
        try {
            // todo: this getting stuff from GeoOptix needs to live in GeoOptixService class
            const geoOptixRequest = await axios.get(
                `${this.baseUrl}/project-overview-web/water-data-program/sites`,
                {
                    headers: this.headers
                }
            );

            const filteredWells = geoOptixRequest.data.filter((x: any) => new Date(x.CreateDate).getFullYear() <= year);

            return filteredWells
                .map((x: { CanonicalName: any; Description: any; Location: any; }) =>
                    ({ wellRegistrationID: x.CanonicalName, description: x.Description, location: x.Location }));
        }
        catch (err) {
            console.error(err);
            throw new InternalServerError(err.message);
        }
    }

    public async getWellSummary(wellRegistrationID: string): Promise<WellSummaryDto | null> {
        try {
            const geoOptixRequest = await axios.get(
                `${this.baseUrl}/project-overview-web/water-data-program/sites/${wellRegistrationID}`,
                {
                    headers: this.headers,
                    validateStatus: (x: number) => true
                }
            );

            if (geoOptixRequest.status === 404 || geoOptixRequest.status === 204){
                return null;
            }

            return {
                location: geoOptixRequest.data.location,
                wellRegistrationID: wellRegistrationID
            };
        }
        catch (err) {
            console.error(err);
            throw new InternalServerError(err.message);
        }
    }

    public async getSensorSummaries(): Promise<SensorSummaryDto[]> {
        try {
            const geoOptixRequest = await axios.get(
                `${this.baseUrl}/project-overview-web/water-data-program/stations`,
                {
                    headers: this.headers
                }
            );

            return geoOptixRequest.data.map((x: {SiteCanonicalName: string, Name: string, Definition: {sensorType: string}}) =>
                ({wellRegistrationID: x.SiteCanonicalName, sensorName: x.Name, sensorType: SensorTypeMap[x.Definition.sensorType]})
            );

        } catch (err) {
            console.error(err);
            throw new InternalServerError(err.message);
        }
    }

    public async getSensorSummariesCreatedAsOfYear(year: number): Promise<SensorSummaryDto[]> {
        try {
            const geoOptixRequest = await axios.get(
                `${this.baseUrl}/project-overview-web/water-data-program/stations`,
                {
                    headers: this.headers
                }
            );

            const filteredSensors = geoOptixRequest.data.filter((x:any) => new Date(x.CreateDate).getFullYear() <= year);

            return filteredSensors.map((x: {SiteCanonicalName: string, Name: string, Definition: {sensorType: string}}) =>
                ({wellRegistrationID: x.SiteCanonicalName, sensorName: x.Name, sensorType: SensorTypeMap[x.Definition.sensorType]})
            );

        } catch (err) {
            console.error(err);
            throw new InternalServerError(err.message);
        }
    }

    public async getSensorsForWell(wellRegistrationID: string): Promise<SensorSummaryDto[]>{
        let geoOptixRequest;
        try {
            geoOptixRequest = await axios.get(
                `${this.baseUrl}/projects/water-data-program/sites/${wellRegistrationID}/stations`,
                {
                    headers: this.headers,
                    validateStatus: (x: number) => true
                }
            );
        } catch (err){
            console.error(err);
            throw new InternalServerError(err.message);
        }

        if (geoOptixRequest.status === 200){
            return geoOptixRequest.data.map(( x: any) =>({
                wellRegistrationID: wellRegistrationID,
                sensorName: x.CanonicalName,
                sensorType: SensorTypeMap[x.Definition.sensorType]
            }));
        } else if (geoOptixRequest.status === 404){
            return [];
        } else{
            console.error(`GeoOptix returned ${geoOptixRequest.status} for Well ${wellRegistrationID} Sensors`);
            throw new InternalServerError(`GeoOptix returned ${geoOptixRequest.status} for Well ${wellRegistrationID} Sensors`);
        }
    }

    // returns a map to facilitate quick cross-reference operations
    public async getWellsWithSensors() : Promise<Map<string,WellWithSensorSummaryDto>>{
        const sensors = await this.getSensorSummaries();
        const wells = await this.getWellSummaries();

        // add a sensor array to the wells
        const wellsWithSensors: WellWithSensorSummaryDto[] = wells.map(x=> ({
            wellRegistrationID: x.wellRegistrationID,
            description: x.description,
            location: x.location,
            sensors: [],
            inGeoOptix: true
        }));

        // create a Map from the array of wells
        const wellMap: Map<string, WellWithSensorSummaryDto> = new Map();
        wellsWithSensors.forEach(x => wellMap.set(x.wellRegistrationID, x));

        // iterate the sensor, adding each one to its well
        sensors.forEach(x => {
            const well = wellMap.get(x.wellRegistrationID);

            if (!well || !x.sensorType){
                return;
            }

            if (!well.sensors){
                well.sensors = []
            }

            well.sensors.push(x);
        });

        // return the Map for further use
        return wellMap;
    }

    public async getInstallationRecord(wellRegistrationID: string): Promise<any> {
        const installationCollectionRoute =
         `${this.baseUrl}/projects/water-data-program/sites/${wellRegistrationID}/samples`;

        try {
            const collectionResult = await axios.get(installationCollectionRoute, {
                headers: this.headers,
                validateStatus: (x: number) => true
            });

            if (collectionResult.status === 404){
                return [];
            }

            const installationRecordDtos = []

            for (const record of collectionResult.data){
                const installationCanonicalName = record?.CanonicalName;
                if (!installationCollectionRoute){
                    return null;
                }
    
                const installationResult = await axios.get(`${installationCollectionRoute}/${installationCanonicalName}/methods`, {
                    headers: this.headers
                });
    
                const installation = installationResult.data;
    
                const installationRecord = installation[0].MethodInstance.RecordSets[0].Records[0].Fields;
                const sensorRecord = installationRecord.sensor.Records[0]?.Fields;
                const photoRecords = sensorRecord.photos.Records;
      
                const sensorType = sensorRecord["sensor-type"] && (Array.isArray(sensorRecord["sensor-type"]) ? sensorRecord["sensor-type"][0] : sensorRecord["sensor-type"]);
                const continuitySensorModel = sensorRecord["sensor-model-continuity"] && (Array.isArray(sensorRecord["sensor-model-continuity"]) ? sensorRecord["sensor-model-continuity"][0] : sensorRecord["sensor-model-flow"]);
                const flowSensorModel = sensorRecord["sensor-model-flow"] && (Array.isArray(sensorRecord["sensor-model-flow"]) ? sensorRecord["sensor-model-flow"][0] : sensorRecord["sensor-model-flow"]);
                const pressureSensorModel = sensorRecord["sensor-model-pressure"] && (Array.isArray(sensorRecord["sensor-model-pressure"]) ? sensorRecord["sensor-model-pressure"][0] : sensorRecord["sensor-model-pressure"]);
      
                const installationRecordDto = {
                    installationCanonicalName: installationCanonicalName,
                    status: installation[0].Status.Name,
                    date: installationRecord["install-date"],
                    lon: installationRecord["gps-location"]?.geometry.coordinates[0],
                    lat: installationRecord["gps-location"]?.geometry.coordinates[1],
                    flowmeterSerialNumber: sensorRecord["flow-serial-number"],
                    sensorSerialNumber: sensorRecord["sensor-serial-number"],
                    affiliation: installationRecord["installer-affiliation"] && installationRecord["installer-affiliation"][0].toUpperCase(),
                    initials: installationRecord["installer-initials"],
                    sensorType: sensorType,
                    continuitySensorModel: continuitySensorModel,
                    flowSensorModel: flowSensorModel,
                    pressureSensorModel: pressureSensorModel,
                    wellDepth: sensorRecord["well-depth"],
                    installDepth: sensorRecord["install-depth"],
                    cableLength: sensorRecord["cable-length"],
                    waterLevel: sensorRecord["water-level"],
                    photos: photoRecords.map((x: any) => x.Fields.photo)
                }
                installationRecordDtos.push(installationRecordDto);
            }

            return installationRecordDtos;
        } catch(err){
            console.error(err);
            throw new InternalServerError(err.message);
        }
    }

    public async getPhoto(wellRegistrationID: string, installationCanonicalName: string, photoCanonicalName: string) {
        const photoUrl = `${this.baseUrl}/projects/water-data-program/sites/${wellRegistrationID}/samples/${installationCanonicalName}/folders/.methods/files/${photoCanonicalName}/view`

        try {
            const result = await axios.get(photoUrl, {
                headers: {
                    "x-geooptix-token": secrets.GEOOPTIX_API_KEY
                },
                responseType: "arraybuffer",
                validateStatus: (number) => true
            });

            if (result.status !== 200){
                return null;
            }

            return result.data
        } catch(err){
            console.error(err);
            throw new InternalServerError(err.message);
        }
    }

    public async getSearchSuggestions(textToSearch: string) : Promise<SearchSummaryDto[]> {
        //PageSize = -1 ensures we get as many results as Azure will let GeoOptix give us
        const searchUrl = `${this.searchUrl}/suggest/${textToSearch}?pageSize=-1`;

        try {
            const result = await axios.get(searchUrl, {
                headers: this.headers
            })

            var searchLowercase = textToSearch.toLowerCase();
            //GeoOptix search will return results where the Description or the Tags have the string
            //We're only really concerned with the Name and CanonicalName, so do some extra filtering and also make sure it isn't a new type
            return result.data.value
            .filter((x: { Document: { ObjectType: string; Name: string; CanonicalName: string }; }) => GeoOptixObjectTypeToZybachObjectType(x.Document.ObjectType) != "" && (x.Document.Name.toLowerCase().includes(searchLowercase) || x.Document.CanonicalName.toLowerCase().includes(searchLowercase)))
            .map((x: { Document: { ObjectType: string; Name: string; SiteCanonicalName: string; }; }) => 
                ({
                    ObjectType : GeoOptixObjectTypeToZybachObjectType(x.Document.ObjectType),
                    ObjectName : x.Document.Name,
                    WellID : x.Document.SiteCanonicalName
                })
            )
            .sort((a: { ObjectName: string; }, b: { ObjectName: string; }) => {
                if (a.ObjectName < b.ObjectName) {
                    return -1;
                }
                if (a.ObjectName > b.ObjectName) {
                    return  1;
                }
                return 0
            });
        }
        catch(err){
            console.error(err);
            throw new InternalServerError(err.message);
        }
    }
}